src/core/bpffs-program.c

```diff
@@ -0,0 +1,92 @@
+/* SPDX-License-Identifier: LGPL-2.1+ */
+
+#include "bpffs-program.h"
+#include "cgroup-bpf-ctx.h"
+
+DEFINE_PRIVATE_HASH_OPS_WITH_VALUE_DESTRUCTOR(bpf_prog_hash_ops, void, trivial_hash_func, trivial_compare_func, BPFProgram, bpf_program_unref);
+
+static int load_programs(Unit *u, const CGroupContext *c, Hashmap *progs_by_id) {
+        CGroupBPFFsProgram *p;
+        int r;
+
+        assert(u);
+        assert(c);
+        assert(progs_by_id);
+
+        LIST_FOREACH(prog, p, c->bpffs_programs) {
+                _cleanup_(bpf_program_unrefp) BPFProgram *prog = NULL;
+
+                r = bpf_program_new_from_bpffs_path(p->bpffs_path, p->attach_type, &prog);
+                if (r < 0)
+                        return log_unit_error_errno(u, r, "Loading of BPF program %s failed: %m", p->bpffs_path);
+
+                r = hashmap_put(progs_by_id, p, prog);
```

Comment:
hmm, i dodn't grok this earlier, but you indx the hashmap by the pointer to the CGroupBPFFsProgram object, but that might not be safe, as clients can typically change cgroup-related stuff for units during runtime, i.e. this might then be an invalid pointer...

Also, "progs_by_id" is a bit of a misnomer, if it is actually indexed by an object, not by an id...

Reply:
Switched to a hash function and a comparator hashing and comparing by value: https://github.com/systemd/systemd/pull/13496/files#diff-11d3c4581343e0c3d561b27621161777ac95427feb2dd615b3c2f59429e80aefR39-R50

There is a hash map with `BPFProgram *` values and (program ID, attach type) keys. The latter uniquely identify what (program ID) should be attached where (attach type), so I named that pair as `BPFForeignKey` in the most recent version.